maturity-level-two/docs/api-security.md

```diff
@@ -0,0 +1,129 @@
+It is possible to bypass server certificate validation on the client side, but this should never be done in any case!
+
+### IP-Filtering
+IP-Filtering is a system where you validate the incoming IP-address on the server. This can be one item or a list. Preferably, this list is as small as possible. 
+
+Note that this does require a static IP of all the users, and by this validation, still ALL machines behind that public-IP can access the resource.
+
+IP-filtering is done ideally on firewall level/API-Management.
+
+### API Key (Shared Access Key)
+The API-Key is a custom HTTP Header with a specific name. Usually this is called `x-API-key`, but can also be called otherwise if required.
+
+The value of this key should be shared between the user and the server.
+The server / API will validate the incoming request via an `Authentication Filter` and verify the value matches between the user and the stored one on the API.
+
+To increase security you can here also use `Rolling API keys` where you allow 2 keys at one time, and you require all the users to update to the new key periodically. This way, if a APIKey is intercepted, it will only be valid for a certain amount of time. Also, having just the mechanism in place and being able to force the rolling at a specific time can be beneficial to fix unwanted resource access.
+
+If you want to implement this in .NET Core, you can speed up things by using the Arcus NuGet package as described [here](https://webAPI.arcus-azure.net/features/security/auth/shared-access-key).
+
+### Basic Authentication
+Basic authentication is the first level of authentication you can apply to distinct users. How this works is standardized.
+
+The users needs to send an `Authorization` header with value
+```
+Basic {username:password}
+```
+where the token `{username:password}` string is a base64 encoding of the actual value.
+
+Again, the server validates the incoming request with an `Authentication Filter` and defines if the user has access or not.
+
+Note that this solution is still very vulnerable to man in the middle attacks as the key the user sends is every time exactly the same.
+So having a policy to frequently change the password might be a good added value in this system.
+
+### TLS Mutual Authentication
+Having a system with Client certificates is a very good way to authenticate the user and by so make your API more secure.
+
+The TLS Mutual authentication does require the user to have a specific certificate given by preferabbly a specific authority.
+
+When the user sends its request including the client certificate the server can then decrypt the certificate on it's side and validate various properties decided on design phase of the API again in an `Authentication Filter`.
+
+Please keep in mind to use validation properties based on things that cannot be changed in a certificate. By this I think about:
+- Certificate Thumprint
+- Certificate Chain
+- Certificate Serial
+- Validity (Expiration time)
```

Comment:
Expiration time, certificate chain,  are automatically validated by the framework. No need for an additional check

Reply:
Not for client certificates. In Azure Functions for example you need write all of that by yourself.

Reply:
> Not for client certificates. In Azure Functions for example you need write all of that by yourself.
u are right